src/manifest/container.rs

```diff
@@ -75,6 +79,238 @@ pub struct Metadata {
+            0xff => None,
+            x => Some(
+                self.toc
+                    .entry(x as usize)
+                    .expect("previously verified by check_invariants()"),
+            ),
+        }
+    }
+
+    /// Returns this entry's hash, if it has one.
+    pub fn hash(self) -> Option<&'toc sha256::Digest> {
+        match self.raw().hash_idx {
+            0xff => None,
+            x => Some(&self.toc.hashes[x as usize]),
+        }
+    }
+
+    /// Returns an iterator over all of this entry's children.
+    pub fn children(self) -> impl Iterator<Item = TocEntry<'entry, 'toc, M>> {
+        self.toc
+            .entries()
+            .filter(move |e| self.index == e.raw().parent_idx as usize)
+    }
+}
+
+// NOTE: Implemented manually, since derive() would generate incorrect bounds
+// M: Clone and M: Copy on the impls.
+impl<M> Clone for TocEntry<'_, '_, M> {
+    fn clone(&self) -> Self {
+        TocEntry {
+            toc: self.toc,
+            index: self.index,
+            _ph: PhantomData,
+        }
+    }
+}
+
+impl<M> Copy for TocEntry<'_, '_, M> {}
+
+/// The table of contents of a PFM.
+///
+/// The table of contents is encoded as follows:
+/// ```text
+/// struct Toc {
+///   entry_count: u8,
+///   hash_count: u8,
+///   hash_type: u8,
+///   entries: [TocEntry; self.entry_count],
+///   hashes: [Digest; self.hash_count],
+///   table_hash: Digest,
+/// }
+/// ```
+///
+/// The layout of the `TocEntry` type is described in [`TocEntry`]. `Digest` is
+/// a hash specified by `hash_type`. Currently, Manticore does not support hash
+/// types other than SHA-256. See [`HashType`] for more information.
+///
+/// The `entries` represent the actual entries to the table of contents; each
+/// entry refers to an *element* in the body of the PFM, describing where it is
+/// and how to decode it.
+///
+/// The `hashes` are digests of certain elements, which entries in the table of
+/// contents refer to. When an element is read out of flash, its hash is
+/// verified against the one present in the TOC, if its TOC entry indicates a
+/// hash. This ensures that the element retains protection of the overall
+/// manifest signature, even if it is parsed long after the TOC has been read
+/// out of flash.
+///
+/// It is not possible to construct a `Toc` directly; it must be parsed out of
+/// a valid [`Pfm`].
+///
+/// [`TocEntry`]: struct.TocEntry.html
+/// [`HashType`]: enum.HashType.html
+/// [`Pfm`]: struct.Pfm.html
+pub struct Toc<'toc, M> {
+    entries: &'toc [RawTocEntry],
+    hashes: &'toc [sha256::Digest],
+    _ph: PhantomData<fn() -> M>,
+}
+
+impl<'toc, M: Manifest> Toc<'toc, M> {
+    /// Checks that all invariants of this `Toc` type hold:
+    /// - Every pointer to a hash is in-bounds.
+    /// - Every pointer to a parent is in-bounds.
+    /// - Every type/version pair is well-known.
+    ///
+    /// Returns true if the invariants are upheld.
+    #[allow(unused)]
```

Comment:
Why do you need to `#[allow(unused)]` here, shouldn't this be being used to check the invariants?

Reply:
Nope.